File: javascript/leetcode/155. Min Stack.js
/**
 * initialize your data structure here.
 */
function Node(x) {
  this.next = null;
  this.data = x;
}

var MinStack = function() {
  this.head = null;
  this.min = null;
};

/**
 * @param {number} x
 * @return {void}
 */
MinStack.prototype.push = function(x) {
  let latest = new Node(x);
  latest.next = this.head;
  this.head = latest;

  let late = new Node(x);
  if (!this.min) this.min = late;
  else {
    if (x <= this.min.data) {
      late.next = this.min;
      this.min = late;
    }
  }
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
  if (this.head) {
    let del = this.head.data;
    this.head = this.head.next;
    if (this.min.data === del) this.min = this.min.next;
  }
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
  return this.head.data;
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
  return this.min.data;
};

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(x)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: c++/leetcode/155. Min Stack.cpp
class MinStack
{
  struct Node
  {
    int data;
    Node *next;
    Node(int val) : data{val} {};
  } * head, *min;

public:
  /** initialize your data structure here. */
  MinStack()
  {
    head = nullptr;
    min = nullptr;
  }

  void push(int x)
  {
    Node *latest = new Node(x);
    latest->next = head;
    head = latest;

    Node *late = new Node(x);
    // add to min
    if (!min)
      min = latest;
    else
    {
      if (x <= min->data)
      {
        late->next = min;
        min = late;
      }
    }
  }

  void pop()
  {
    if (head)
    {
      int del = head->data;
      head = head->next;
      if (min->data == del)
      {
        min = min->next;
      }
    }
  }

  int top()
  {
    return head->data;
  }

  int getMin()
  {
    return min->data;
  }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */